File: Java/MergeSort.java
import java.util.Arrays;

/*
Time: O(nlogn), Space: O(n)
*/
public class MergeSort {

    public static void sort(int[] a, int l, int r) {
        if (l < r) {
            int pvtIdx = (l + r) / 2;
            sort(a, l, pvtIdx);
            sort(a, pvtIdx + 1, r);
            merge(a, l, pvtIdx, r);
        }
    }

    private static void merge(int[] a, int l, int m, int r) {
        int[] a1 = new int[m - l + 1]; //[l..m]
        int[] a2 = new int[r - m]; // [m+1..r]
        for (int i = 0; i < a1.length; i++) {
            a1[i] = a[l + i];
        }
        for (int i = 0; i < a2.length; i++) {
            a2[i] = a[m + 1 + i];
        }
        
        int i = 0, j = 0, k = l;
        while (i < a1.length && j < a2.length) {
            if (a1[i] <= a2[j]) {
                a[k++] = a1[i++];
            } else {
                a[k++] = a2[j++];
            }
        }

        while (i < a1.length) {
            a[k++] = a1[i++];
        }

        while (j < a2.length) {
            a[k++] = a2[j++];
        }
    }

    public static void main(String[] args) {
        int[] a = new int[] { 2, 7, 1, 6 };
        sort(a, 0, a.length - 1);
        System.out.println("Test case 1: {2,7,1,6} " + Arrays.equals(a, new int[] { 1, 2, 6, 7 }));
        a = new int[] { 4, 3, 2, 1 };
        sort(a, 0, a.length - 1);
        System.out.println("Test case 2: {4,3,2,1} " + Arrays.equals(a, new int[] { 1, 2, 3, 4 }));
        a = new int[] { 1, 2, 3, 4 };
        sort(a, 0, a.length - 1);
        System.out.println("Test case 3: {1,2,3,4} " + Arrays.equals(a, new int[] { 1, 2, 3, 4 }));
    }

}

File: Python/MergeSort.py
def merge_sort(a, l, r):
    if l < r:
        pvtIdx = (l + r) // 2
        merge_sort(a, l, pvtIdx)
        merge_sort(a, pvtIdx + 1, r)
        merge(a, l, pvtIdx, r)

def merge(a, l, m, r):
    a1 = a[l:m+1]
    a2 = a[m+1:r+1]
    
    i = j = 0
    k = l
    while i < len(a1) and j < len(a2):
        if a1[i] <= a2[j]:
            a[k] = a1[i]
            i += 1
        else:
            a[k] = a2[j]
            j += 1
        k += 1
    
    while i < len(a1):
        a[k] = a1[i]
        i += 1
        k += 1
    
    while j < len(a2):
        a[k] = a2[j]
        j += 1
        k += 1

a = [2, 7, 1, 6]
merge_sort(a, 0, len(a) - 1)
print(f"Test case 1: {a} {a == [1, 2, 6, 7]}")

a = [4, 3, 2, 1]
merge_sort(a, 0, len(a) - 1)
print(f"Test case 2: {a} {a == [1, 2, 3, 4]}")

a = [1, 2, 3, 4]
merge_sort(a, 0, len(a) - 1)
print(f"Test case 3: {a} {a == [1, 2, 3, 4]}")